### ISSUE TITLE
No Error Message - using pyodbc.connect()

### ISSUE THREAD

**Issue by Bigbigyoshi** [mkleehammer/pyodbc#608]
### Environment
- Python: 3.6.6 64bit
- pyodbc: 4.0.27
- OS:  Windows 10 Enterprise
- DB: Microsoft Access DB Linked to Sharepoint 2016 List
- driver: Not sure how to check this. Please advise if this info is important.

### Issue
I have two .accdb files. One is called Test.accdb, the other is called Test2.accdb.

Test.accdb is **linked to a Sharepoint 2016 List.**
Test2.accdb is **not linked to anything**.

I am having issues setting up a `conn` variable. 

I am assigning `conn` as follows:
```
conn = pyodbc.connect(r'Driver={Microsoft Access Driver (*.mdb, *.accdb)}; DBQ=C:\\Users\\userName\\Desktop\\Project Folders\\sharepointList\\test.accdb;')
```

When I run that line of code with test2.accdb as the file to connect to I have no issues. The line completes and I can run the rest of my code.

When I run the line of code with test.accdb, the one **linked to Sharepoint,** I can not get past it. 
I wrote a small amount of code to test a theory. It is shown below:
```
print("Before")
conn = pyodbc.connect(r'Driver={Microsoft Access Driver (*.mdb, *.accdb)}; DBQ=C:\\Users\\userName\\Desktop\\Project Folders\\sharepointList\\test.accdb;')
print("After")
```

I also tried the following, using try/except:
```
try:
    print("Before")
    conn = pyodbc.connect(r'Driver={Microsoft Access Driver (*.mdb, *.accdb)}; DBQ=C:\\Users\\userName\\Desktop\\Project Folders\\sharepointList\\test.accdb;')
except:
    print("Something Crashed")
print("After")
```
The new print statement was never executed. This leads me to believe that the code never really crashed, It just sort of stopped. 

When I run this code with test2.accdb both print lines are executed.
When I run this code with test.accdb, the one **linked to sharepoint**, only the "Before" print line is executed. 
I would expect the `conn` assignment to execute completely and allow me to execute the rest of my code. Instead, the code stops running at some point within the .connect() function.
**I get no error message when I run the code. The code simply stops.**

**Comment by gordthompson** [comment]
Have you tried executing `pyodbc.pooling = False` before calling `connect` as suggested [here](https://github.com/mkleehammer/pyodbc/wiki/Tips-and-Tricks-by-Database-Platform#databases-containing-odbc-linked-tables)?

**Comment by Bigbigyoshi** [comment]
> Have you tried executing `pyodbc.pooling = False` before calling `connect` as suggested [here](https://github.com/mkleehammer/pyodbc/wiki/Tips-and-Tricks-by-Database-Platform#databases-containing-odbc-linked-tables)?

I have tried adding 'pyodbc.pooling = False' directly before calling `connect`. I got the same results.

**Comment by v-chojas** [comment]
Could you post an ODBC trace?

**Comment by Bigbigyoshi** [comment]
> Could you post an ODBC trace?

Where can I find this?

**Comment by v-chojas** [comment]
[Instructions here](https://docs.microsoft.com/en-us/sql/odbc/admin/setting-tracing-options)

**Comment by Bigbigyoshi** [comment]
Start the trace, run my script, end the trace?

**Comment by v-chojas** [comment]
Correct.

**Comment by gordthompson** [comment]
It might also be helpful to know if the following VBScript works for you:

```vbscript
driver = "{Microsoft Access Driver (*.mdb, *.accdb)}"
dbq = "C:\Users\userName\Desktop\Project Folders\sharepointList\test.accdb"
Set con = CreateObject("ADODB.Connection")
con.Open "DRIVER=" & driver & ";DBQ=" & dbq
WScript.Echo "Connected."
```

**Comment by Bigbigyoshi** [comment]
I have found some trace information in an SQL.LOG file. There are 7 `temp` readouts that have appeared after running the trace around my failing script. Three of them have return codes of 0, or successful, the other 4 do not have any return codes.

**Comment by Bigbigyoshi** [comment]
> driver = "{Microsoft Access Driver (*.mdb, *.accdb)}"
> dbq = "C:\Users\userName\Desktop\Project Folders\sharepointList\test.accdb"
> Set con = CreateObject("ADODB.Connection")
> con.Open "DRIVER=" & driver & ";DBQ=" & dbq
> WScript.Echo "Connected."

In attempting to run this script in powershell, Powershell 'stopped working'.

**Comment by v-chojas** [comment]
That is not a Powershell script, it is a VBscript.

**Comment by Bigbigyoshi** [comment]
> That is not a Powershell script, it is a VBscript.

Can I not just run the script in CMD or Powershell?

**Comment by gordthompson** [comment]
No, VBScripts need to be interpreted by CSCRIPT.EXE or WSCRIPT.EXE.

The equivalent PowerShell script would be

```powershell
$connStr = @"
DRIVER={Microsoft Access Driver (*.mdb, *.accdb)};
DBQ=C:\Users\userName\Desktop\Project Folders\sharepointList\test.accdb
"@
$con = New-Object System.Data.Odbc.OdbcConnection $connStr
$con.Open()
Write ("Connected.")
```

**Comment by Bigbigyoshi** [comment]
Same thing happens when I run the .ps1 as when I ran the VB Script using cscript(path to script): I got a pop-up telling me that PowerShell ISE has stopped working. Nothing was printed in the console as far as error codes go.

**Comment by gordthompson** [comment]
It's beginning to sound like the Access ODBC driver simply doesn't like that .accdb file.

**Comment by gordthompson** [comment]
... and if that is indeed the case then you might have better luck using [pywin32](https://pypi.org/project/pywin32/) and ACE.DAO instead of pyodbc and ODBC:

```python
import win32com.client

# ACE.DAO constants
dbOpenSnapshot = 4

dbe = win32com.client.Dispatch("DAO.DBEngine.120")
db = dbe.OpenDatabase(r"C:\Users\Public\Database1.accdb")
rs = db.OpenRecordset("SELECT DonorName FROM Donor WHERE DonorID=1", dbOpenSnapshot)
print(rs.Fields["DonorName"].Value)
rs.Close()
db.Close()
```

**Comment by Bigbigyoshi** [comment]
Strangely enough I get the same sort of issue when using this, only, it happens with the line:
`db = dbe.OpenDatabase(r"test.accdb")`. 

Are there limitations with using these methods on a .accdb that's linked to sharepoint by any chance?

**Comment by gordthompson** [comment]
If you open the database in Access itself are you prompted for credentials when you try to open an object that is linked to SharePoint?

**Comment by Bigbigyoshi** [comment]
> If you open the database in Access itself are you prompted for credentials when you try to open an object that is linked to SharePoint?

No. When I open the .accdb I can access everything. I can also add items to the .accdb and have them successfully transfer to the sharepoint list.  Something to note though, I am using Microsoft Office 365. The account I am signed into there has access to the Sharepoint. If I was not signed into the account for Office 365 I might not be able to access the .accdb. I would think that since the .accdb can already talk to sharepoint I wouldn't need any login info for the pyodbc script.

**Comment by v-chojas** [comment]
This might be a limitation of the Access ODBC driver.

**Comment by gordthompson** [comment]
You don't have the source files for the ACE.ADO components. Only Microsoft has those.

One other thing you could try, again with pywin32:

```python
import win32com.client

# ACE.DAO constants
dbOpenSnapshot = 4

print('Launching MSACCESS.EXE in the background ...')
appAccess = win32com.client.Dispatch("Access.Application")
print('... done.')
print('Opening database ...')
appAccess.OpenCurrentDatabase(r"C:\Users\Public\Database1.accdb")
print('... done.')
cdb = appAccess.CurrentDb()
rs = cdb.OpenRecordset("SELECT txt FROM Table1 WHERE ID=1", dbOpenSnapshot)
print(rs.Fields["txt"].Value)
rs.Close()
cdb.Close()
appAccess.Quit()
```

**Comment by Bigbigyoshi** [comment]
I'm not entirely sure what should be used in place of dbOpenSnapshot. Is this something that should be left as is? Currently it is being treated as a variable that has not been declared.

**Comment by gordthompson** [comment]
Have your test script `print(os.getcwd())` to verify that the current working directory is in fact the directory where "test.accdb" resides.

**Comment by Bigbigyoshi** [comment]
> Have your test script `print(os.getcwd())` to verify that the current working directory is in fact the directory where "test.accdb" resides.

Yes. I also confirmed by testing test2.accdb with this script. (The file not liked to sharepoint)
Everything worked when I used test2.accdb and I have confirmed that my path to the file is correct.

**Comment by gordthompson** [comment]
Well, just on the off-chance that it is some incompatibility between Python and that particular file you could always try this in PowerShell:

```powershell
Write "Launching MSACCESS.EXE in the background ..."
$Access = New-Object -com Access.Application
Write "... done."
Write "Opening database ..."
$Access.OpenCurrentDatabase("C:\path\to\test.accdb")
Write "... done."
Write "Closing database ..."
$Access.CloseCurrentDatabase()
Write "... done."
Write "Quitting MSACCESS.EXE ..."
$Access.Quit()
Write "... done."
$rtn = [System.Runtime.InteropServices.Marshal]::ReleaseComObject($Access)
Remove-Variable Access
```

**Comment by Bigbigyoshi** [comment]
Seeing a successful run is a sight for sore eyes. This PowerShell script seemed to do the trick.
```
PS C:\Users\UserName\Desktop> C:\Users\UserName\Desktop\powershellscript.ps1
Launching MSACCESS.EXE in the background ...
... done.
Opening database ...
... done.
Closing database ...
... done.
Quitting MSACCESS.EXE ...
... done.
```

**Comment by gordthompson** [comment]
>  it would be best if this project could be kept in python

That could prove to be ... "challenging". Everything we've tried in Python has been unsuccessful, including COM Automation (`Access.Application`), but this last test indicates that COM Automation works (i.e., can at least open the linked database without crashing) from PowerShell. Unfortunately I was not able to get ACE.DAO to work from within PowerShell, even for a "normal" Access database, and depending on what you actually want to accomplish PowerShell might not be the best environment anyway.

If this was my project my next step would be to fire up Visual Studio and see if I could work with the linked database via COM Automation from C#.

**Comment by Bigbigyoshi** [comment]
> > it would be best if this project could be kept in python
> 
> That could prove to be ... "challenging". Everything we've tried in Python has been unsuccessful, including COM Automation (`Access.Application`), but this last test indicates that COM Automation works (i.e., can at least open the linked database without crashing) from PowerShell. Unfortunately I was not able to get ACE.DAO to work from within PowerShell, even for a "normal" Access database, and depending on what you actually want to accomplish PowerShell might not be the best environment anyway.
> 
> If this was my project my next step would be to fire up Visual Studio and see if I could work with the linked database via COM Automation from C#.

That's a bummer that we couldn't find a solution using Python. 
Thanks a ton for all the help and time you've provided!

**Comment by gordthompson** [comment]
This issue could do with a more descriptive title that mentions Access and SharePoint.
